Add visual transposition for keyboard and renderer

spessasus opened the pull request:
This PR adds visual transposition for the keyboard and renderer, shifting the notes accordingly.


@TS53, is this the result you expect? Let me know :-)

Note that the mouse note presses also get shifted to show the user how the notes are shifted. This is intentional

Fixes #128

TS53 commented:
This looks really good, thank you!
I'm not sure about "mouse note presses", I only have a notebook without mouse.

Here are the things I noticed:
- when setting the keyboard size from 128 to 5 octaves it is easy to completely hide the piano keys. While this is correct, it could be more user-friendly to still show the notes in the last valid visible octave that is currently visible on the screen. If at least one note is not visible anymore, show the last valid octave. But I'm not sure about this. Maybe you can also let it the current way.
- only sometimes at the beginning the changes are immediately visible in pause mode, but then the visual notes will only update when the midi continues playing
- On small screens, the synthesizer controller covers the complete window, to also see the keys at the same time, I would have to press "Ctrl" + "-" to zoom smaller. Maybe it would be good to only show the midi channels that are available in the MIDI file which would improve for smaller screens.

spessasus commented:
Thanks again for your feedback!

By mouse, I just meant the cursor clicking on the virtual keyboard. I have a laptop too and I don't use a mouse :-) It means that when you press a key and the channel is transposed, the actual key you press will be shifted.

Now, addressing your feedback:
- I would suggest against lighting up the highest note if it is out of range, as that would be confusing and technically incorrect. The user would see a key getting pressed and a different note being played, confusing them as to why the audio is playing several octaves lower. Not to mention that doing this would cause all the instruments to play in the max octave, even if in the real MIDI file they are shifted by let's say, two octaves.
- That's because the renderer does not render when nothing is playing and the sequence is paused. But you're right, missed that. I'll fix it once I get back home.
- the synthesizer controller is intended to cover up the entirety of the screen below the top part. On my laptop it always reaches all the way to the bottom. In fact, [try playing one of these MIDI files](https://github.com/user-attachments/files/16216556/multi_port_collection.zip). You'll see that it can become even larger!

TS53 commented:
Now I see what you mean! Yes, this works good!
I can understand the decision regarding the shifting of keys.
Regarding the limiting of the midi channels, I hoped that I could see the piano keyboard while transposing the channel. I could then immediately see on the keyboard how it effects the notes. Currently, this works by zooming out very much so the piano appears. If this was even working in paused mode, it would be even simpler to see the changes. During the playback the notes change which makes it more difficult to see the result.

spessasus commented:
I think I've found a nice solution to that problem: When you change the tranpose controller, all the controllers below it disappear so you can see the keyboard!

TS53 commented:
The solution is indeed very good. But it introduces a small bug: when opening the synthesizer controller, there is a window animation while it is not 100 % opened. But now, during this animation, the keyboard, piano notes and player menu flicker trough. It seems to be a layer problem, but not sure. When it is opened completely, the flickering goes away. This also happens with no midi file open.

spessasus commented:
It happened to me before this change. It looks like a Firefox bug because it does not happened on chromium.

TS53 commented:
Ok, I have not seen it before, but I am using Firefox. Maybe this was there before.

TS53 commented:
It seems to happen only when either no midi is loaded, or when a midi is loaded and in pause mode.

TS53 commented:
Yes, also settings.

One small issue: it only hides all channels above. So when I wanted to transpose channel number 10, the problem is the same again. I think instead of hiding all above, it would be better to hide all others beside the transposed, if possible.
This will also not cover all cases, but it will cover ~90% of all channel cases, much more than now.

spessasus commented:
I don't think I can do that. And a correction, it hides all the channels below. Hiding channels uses `display: none` which removes them from the layout. If I did remove all channels above the currently transposed one, it would immediately move to the top (1st channel's spot) which would make the cursor be out of range, stopping the transposition from happening...

TS53 commented:
Maybe temporarily setting the zindex of these channels to the lowest or something. I assume it would be possible to hide them without changing the order. But if it turns out difficult, it is no problem. It is already very good now, thank you!
